### PR TITLE
Experiment: commit whitespace changes before computing indentation.

### PR DESCRIPTION
**Pull request by matklad**
The biggest issue with our current formatting infra is that patterns
are evaluated on the original syntax tree, and not on the formatting
model, which might contain whitespace changes.

A particular bad issue here is when whitespace changes add or remove
newlines, but then we use newline info to calculate indent.
Originally, the thinking was that this is not necessary, as rules
would only depend on the shape of the tree and not on the spacing, but
turns out this is not the case in practice.

The "obvious" fix here is to make our formatting model & predicated
more complicated, so that they can take modified whitespace into
account. But this PR proposes something more stupid, yet more
elegant (subjectively) -- we just commit whitespace changes to the
syntax tree before re-indent pass.

That is, we first fix whitespace, then modify the syntax tree to take
updated ws into account (by using to_string and reparse), and then
re-indent the new tree.

This will be a bit slower, as we have one extra step of tree
re-building, but the perf should not matter *that* much (rust is not a
slow language), and we can always make the internal model more
perofrmant and zero-copy later.

### CODE REVIEW

File: src/rules.rs [hunk 1]
```diff
@@ -107,10 +107,10 @@ pub(crate) fn spacing() -> SpacingDsl {
         .inside(NODE_LET_IN).before(T![in]).when(header_has_multi_value).when(before_token_not_newline).newline()
         .inside(NODE_LET_IN).after(T![in]).when(in_body_newline).newline()
         .inside(NODE_LET_IN).before(NODE_KEY_VALUE).when(header_has_multi_value).when(before_token_not_newline).newline()
-        
+
         .test("{a?3}: a", "{ a ? 3 }: a")
         .inside(NODE_PAT_ENTRY).around(T![?]).single_space()
-
```
Comment (matklad) [inline src/rules.rs:113]: @Rizary you might want to tweak your editor to strip whitespace on save. I think in Emacs this is handled by whitespace cleanup mode: https://github.com/purcell/whitespace-cleanup-mode

File: src/rules.rs [hunk 2]
```diff
@@ -748,18 +748,10 @@ pub(crate) fn indentation() -> IndentDsl {
...
             .inside(p(NODE_LAMBDA) & p(not_on_top_level) & p(pattern_not_newline))
             .set(Indent)
-        .rule("Indent newline lambda body")
-            .inside(p(NODE_LAMBDA) & p(not_on_top_level) & p(pattern_newline) & p(lambda_inside_node_pattern))
-            .not_matching(p(TOKEN_COMMENT)) 
-            .set(Indent)
-         .rule("Indent newline lambda body")
-            .inside(p(NODE_LAMBDA) & p(not_on_top_level) & p(pattern_newline) & p(lambda_outside_node_pattern))
-            .not_matching(p(TOKEN_COMMENT) | p(VALUES)) 
-            .set(Indent)
```
Comment (matklad) [inline src/rules.rs:762]: Haven't check other rules, but these two seem unnecessary with the change to the engine.

File: src/rules.rs (hunk 1 above)
Comment (Rizary) [reply src/rules.rs:113]: @matklad I think I have it enable. Let me check with my doom emacs configuration.

File: src/rules.rs (hunk 2 above)
Comment (Rizary) [reply src/rules.rs:762]: Beside the `explain_smoke_test` failure`, I found two other rules that can be removed which are regarding node binops and inline node if_else.

**Comment by matklad** [comment]
Ok, I think this is now ready to be merged @Rizary !

I really like this new model, seems much more intuitive to work this. Should have done this from the start, but this only became obvious after stepping from the code for a while :)

@Rizary I think I've accidentally stomped over your commits here, would you mind r+ this PR and then applying your changes to master? In general, commiting together to the same branch is difficult coordination-wise, so imo (but this is very subjective) it's better to open separate feature branches or just dump interesting diffs/suggestions as comments.

**Comment by Rizary** [comment]
@matklad sure. 

I'll push my fixed in my next PR.

bors r+

**Pull request merged**